Demo/MozillaSocial-iOS/MoSoContent/AuthenticationService.swift

```diff
@@ -0,0 +1,207 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import AuthenticationServices
+import Combine
+
+class AuthenticationService: NSObject, ASWebAuthenticationPresentationContextProviding, ObservableObject {
+
+    @Published var authToken: Token?
+    @Published var accountDetails: AccountDetails?
+
+    private var subscribers: Set<AnyCancellable> = []
+
+    private enum URLConstants {
+        static let isProd = false
+        static let baseURL = URLConstants.isProd ? "https://mozilla.social" : "https://stage.moztodon.nonprod.webservices.mozgcp.net"
+        static let redirectScheme = "mozillasocial"
+        static let redirectURI = URLConstants.redirectScheme + "://auth"
+    }
+
+    init(authToken: Token? = nil, accountDetails: AccountDetails? = nil) {
+        super.init()
+
+        self.authToken = authToken
+        self.accountDetails = accountDetails
+
+        $authToken.sink { recievedValue in
+            guard let bearerToken = recievedValue?.accessToken else {
+                return
+            }
+            Task {
+                let accountDetails = await self.fetchAccountDetails(for: bearerToken)
+                await MainActor.run {
+                    self.accountDetails = accountDetails
+                }
+            }
+        }
+        .store(in: &subscribers)
+    }
+
+    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+        return ASPresentationAnchor()
```

Comment:
I was getting a main thread checker runtime error here, saying that a UI api was called from a background thread: maybe worth checking it?

Reply:
Funny, XCode asked for that, then I did a refactor and removed it and it didn't complain.